### PR TITLE
Isaac Del Rosario -- Carets

### PR DESCRIPTION
**Pull request by hisaacdelr**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a set_default method that sets the creator to Anonymous if the user doesn't enter in a creator value for a work.
Describe how you approached testing that model method. What edge cases did you come up with? | Admittedly, I didn't test for this method (which I realized later could be done by setting defaults in the model w/o a custom method. I barely had custom model methods and couldn't think of other instances when I can create a method for it.
Describe an edge case test you wrote for a controller | If a user tries to access a non-valid work show page, i.e. a show page of a work that doesn't exist, I test if the user is redirected to the 404 page
What are `session` and `flash`? What is the difference between them? | Session is used for persisting information such as log-in information and these are stored in cookies in the browser. Flash is used for showing non-persisting messages to the user that you only want to display once a new view is created.
Describe a controller filter you wrote. | For my upvote function, I check if a) a user is not logged in, b) a user is upvoting for the first time on a work, and c) if a user is trying to upvote again on a work. For a., I show a flash message to prompt the user to log in, for b., being the optimal route, I create a new Vote object with the work and the currently logged-in user references and save it to the database. And for c., I show a flash message that says the user already voted for this work
What was one thing that you gained more clarity on through this assignment? | using custom routes and methods in my controller. Admittedly a little bit on testing because I did not do the assignment with a test-driven approach. But rather a "make sure it works with tests" approach by creating tests afterward.
What is the Heroku URL of your deployed application | http://media-ranker-c8-isaacdelr.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | Not specific to the project but going over how to approach doing test cases together would have helped with my approach to testing for the project. Testing was very overwhelming in general of a subject for me. Especially since realizing that our previous lectures' codes for ada books built upon each other and if I couldn't implement a section, then I can't move on (ex. testing for sessions when I struggled a bit with how sessions worked). Given more time, I'm now confident I could get session testing to work.

### CODE REVIEW

File: app/models/work.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+                          scope: :category,
+                          message: "This work already exists"
+                        }
+  validates :publication_year,
+            allow_blank: true,
+            numericality: {
+              only_integer: true,
+              greater_than_or_equal_to: 1000
+            }
+
```
Comment (CheezItMan) [inline app/models/work.rb:19]: You should have a method that returns the list of works by category ordered by the number of votes.

**Comment by CheezItMan** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | Good number of commits and good commit messages. |
Comprehension questions	| Check |
**General** |
Rails fundamentals (RESTful routing, use of named paths) | Check, well done |
Semantic HTML | You can use things like:  `class="row"` in a `header` or other semantic element.  That might help DRY up some of the excess `div` elements.  |
Errors are reported to the user |  You did need improved styling on errors, so they look different from confirmation messages.  Also things like show actions for works that don't exist crash the app.  
Business logic lives in the models |  Not very much here.  You should have things like spotlight and sorting of works by votes within a category.    |
Models are thoroughly tested, including relations, validations and any custom logic |  No real tests for `Vote` or `User`.  You did ok at testing validations for Work, and you should test both ends of the creator custom method.  |
Controllers are thoroughly tested | No testing for creating a User's create action.  You also need more negative tests for the Works controller (like deleting a work that doesn't exist, or showing an item that isn't there).   |
**Wave 1 - Media** |
Splash page shows the three media categories |  Check
Basic CRUD operations on media are present and functional | Check
**Wave 2 - Users and Votes** |
Users can log in and log out | Check |
The ID of the current user is stored in the session |  Check |
Individual user pages and the user list are present | Check |
A user cannot vote for the same media more than once | Check |
All media lists are ordered by vote count | Check, on my localhost.  Heroku doesn't have that working.   |
Splash page contains a media spotlight |  Check |
Media pages contain lists of voting users |  Check |
**Wave 3 - Styling** |
Foundation is used appropriately |  Good work here, decent responsiveness.  The menus look a little weird. |
Look and feel is similar to the original |  Pretty close.  |
**Overall** |  Not bad, you've got a lot of work to do with putting business logic in the model and writing good tests.  In bEtsy  and Media-Ranker-OAuth, I'd like you to make a focused effort to doing good tests.  You can use the the tests already in Media-Ranker-OAuth as examples.  Overall well done, just focus more on learning testing and putting logic in the models.    |
